File: solutions/ruby/kindergarten-garden/1/kindergarten_garden.rb
class Garden
  ROW_COUNT     = 2
  ROW_SEPARATOR = "\n"

  PLANT_CODES = {
    "G": "grass",
    "C": "clover",
    "R": "radishes",
    "V": "violets"
  }.freeze

  STUDENTS = [
    :alice,
    :bob,
    :charlie,
    :david,
    :eve,
    :fred,
    :ginny,
    :harriet,
    :ileana,
    :joseph,
    :kincaid,
    :larry
  ].freeze

  def initialize(diagram)
    @window_rows = diagram.split(ROW_SEPARATOR)
  end

  def method_missing(m, *args, &block)
    plants(ROW_COUNT * STUDENTS.index(m.to_sym))
  end

  private

  def plants(start)
    codes     = []
    locations = (0...ROW_COUNT).map { |i| start + i }

    @window_rows.each do |row|
      locations.each { |i| codes << row.chars[i] }
    end

    codes.map { |c| PLANT_CODES.fetch(c.to_sym).to_sym }
  end
end

File: solutions/javascript/kindergarten-garden/1/kindergarten-garden.js
const PLANTS_PER_STUDENT   = 2;
const WINDOW_SEPARATOR     = '\n';
const PLANT_CODE_SEPARATOR = '';

const DEFAULT_STUDENTS = [
  'Alice',
  'Bob',
  'Charlie',
  'David',
  'Eve',
  'Fred',
  'Ginny',
  'Harriet',
  'Ileana',
  'Joseph',
  'Kincaid',
  'Larry',
];

const PLANT_CODES = {
  G: 'grass',
  V: 'violets',
  R: 'radishes',
  C: 'clover',
};

export class Garden {
  #windowRows = [];
  #students   = [];

  constructor(diagram, students = DEFAULT_STUDENTS) {
    this.#windowRows = diagram.split(WINDOW_SEPARATOR);
    this.#students   = students.sort();
  }

  plants(student) {
    const startIndex = this.#students.indexOf(student) * PLANTS_PER_STUDENT;

    let locations  = [startIndex];
    let plantCodes = [];

    for (let i = 1; i < PLANTS_PER_STUDENT; i++) {
      locations.push(startIndex + i);
    }

    this.#windowRows.forEach(row => {
      const rowPlantCodes = row.split(PLANT_CODE_SEPARATOR);

      for (let i = 0; i < locations.length; i++) {
        plantCodes.push(rowPlantCodes[locations[i]]);
      }
    });

    return plantCodes.map(c => PLANT_CODES[c]);
  }
}
